### PR TITLE
firefox-*: change binary name and desktop entry

### PR DESCRIPTION
**Pull request by kxxt**
Firefox and its other editions shouldn't have the same binary name and desktop entry because it makes it hard to use two of these browsers at the same time.

For example, In a common use case, one use Firefox for life and use Firefox Dev Edition for web development.

This also makes the `firefox-*` packages more consistent with the `firefox-*-bin` packages.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kxxt** [comment]
Converted to a draft. Let's change firefox-beta as well for consistency.

**Review by bachp — APPROVED** [review]

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -30812,11 +30812,23 @@ with pkgs;
...
+    desktopName = "Firefox DevEdition";
+    wmClass = "firefox-devedition";
+  };
 
   firefox-esr = firefox-esr-102;
-  firefox-esr-102 = wrapFirefox firefox-esr-102-unwrapped { };
+  firefox-esr-102 = wrapFirefox firefox-esr-102-unwrapped {
+    nameSuffix = "-esr";
+    desktopName = "Firefox ESR";
+    wmClass = "firefox-esr";
```
Comment (mweinelt) [inline pkgs/top-level/all-packages.nix:30830]: We usually have two ESR releases for 3 months or so every once in a while, when they switch between one and the next.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -30812,11 +30812,23 @@ with pkgs;
...
-  firefox-beta = wrapFirefox firefox-beta-unwrapped { };
-  firefox-devedition = wrapFirefox firefox-devedition-unwrapped { };
+  firefox-beta = wrapFirefox firefox-beta-unwrapped {
+    nameSuffix = "-beta";
+    desktopName = "Firefox Beta";
+    wmClass = "firefox-beta";
+  };
+  firefox-devedition = wrapFirefox firefox-devedition-unwrapped {
+    nameSuffix = "-devedition";
+    desktopName = "Firefox DevEdition";
```
Comment (mweinelt) [inline pkgs/top-level/all-packages.nix:30822]:
```suggestion
    desktopName = "Firefox Developer Edition";
```

https://www.mozilla.org/de/firefox/developer/

Comment (kxxt) [reply pkgs/top-level/all-packages.nix]: Corrected

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (kxxt) [reply pkgs/top-level/all-packages.nix]: I think the best way to handle this is to make firefox-esr a symlink to firefox-esr-102 but it is not obvious to me about how to do this in nix.

Maybe we can leave it for later because it will not be a breaking change.

**Comment by kxxt** [comment]
I updated the icon names so that we won't end up having the same icon for firefox and firefox-devedition.

File: nixos/doc/manual/release-notes/rl-2311.section.md [hunk 3]
```diff
@@ -36,6 +36,8 @@
 
 - `fileSystems.<name>.autoResize` now uses `systemd-growfs` to resize the file system online in stage 2. This means that `f2fs` and `ext2` can no longer be auto resized, while `xfs` and `btrfs` now can be.
 
+- The executable file names for `firefox-devedition`, `firefox-beta`, `firefox-esr` now matches their package names, which is consistent with the `firefox-*-bin` packages. The desktop entries are also updated so that you can have mulitple editions of firefox in your app launcher.
+
```
Comment (lafrenierejm) [inline nixos/doc/manual/release-notes/rl-2311.section.md:40]: This would need to be moved to a different release now that 23.11 is out. Might be best to just leave release notes for a separate PR?

**Review by lafrenierejm — CHANGES_REQUESTED** [review]

File: nixos/doc/manual/release-notes/rl-2311.section.md (hunk 3 above)
Comment (kxxt) [reply nixos/doc/manual/release-notes/rl-2311.section.md]: Done

**Comment by kxxt** [comment]
Hi all, it seems that there are still interest in this PR. So I fixed the merge conflicts. But as I am no longer using NixOS for now, I might close this PR if another merge conflict happens in the future.

**Review by lafrenierejm — APPROVED** [review]
